### PR TITLE
drivers: spi: improve initialization of multiple cs gpios

### PR DESCRIPTION
**Pull request by bbilas**
Initialize all defined cs gpios and set them as inactive during driver initialization to be sure that before first communication over SPI the cs gpio is in the valid state. It also fixes the case when we have multiple devices connected to the one SPI interface and the initial state of cs gpios after MCU reset is floating and might be low, so that prevents us from communicating between particular devices. 

PR has been tested on STM32F4 with 15 TMC5160 stepper motor controllers connected via SPI.

That's an updated version of https://github.com/zephyrproject-rtos/zephyr/pull/36982

### CODE REVIEW

File: drivers/spi/spi_context.h
Comment (gmarull) [inline drivers/spi/spi_context.h:66]: `gpio_dt_spec` (and associated APIs) can be used instead, will make things easier

Comment (bbilas) [reply drivers/spi/spi_context.h:66]: I could have used this structure but It's not compatible with spi_cs_control structure that I'm using here.

Comment (gmarull) [reply drivers/spi/spi_context.h:66]: then, wouldn't it be better to cleanup that part, too?

Comment (mniestroj) [reply drivers/spi/spi_context.h:66]: @gmarull It would. But I see that there are plenty of drivers to be modified to be `gpio_dt_spec` compatible, so I think this is a topic for another PR. Besides, I would probably first convert drivers to use `spi_dt_spec`, so after that the conversion to `gpio_dt_spec` would be probably just a change of single macro (`SPI_CS_CONTROL_PTR_DT`).

Comment (JordanYates) [reply drivers/spi/spi_context.h:66]: FYI I have started converting drivers to `spi_dt_spec` in #37358 and have commits to convert the spi_cs_control to `gpio_dt_spec` here: https://github.com/csiro-wsn/zephyr/commits/210801_spi_cs

File: drivers/spi/spi_ll_stm32.c
Comment (mniestroj) [inline drivers/spi/spi_ll_stm32.c:838]: There is nothing stm32 specific in this function, so I would rather have one generic function that already iterates through all CS gpios.

Comment (bbilas) [reply drivers/spi/spi_ll_stm32.c:838]: I'm aware of that but I wasn't sure if I should create another function like e.g spi_cs_configure_all that will call spi_cs_configure or create the separate functions for each arch SPI driver and call them in the init function. Now I see that a generic approach would be more convenient so I'm gonna fix that.

File: drivers/spi/spi_context.h
Comment (tbursztyka) [reply drivers/spi/spi_context.h:66]: perhaps it would be better to get spi_cs_control to gpio_dt_spec first then, and afterwards this patch

Comment (bbilas) [reply drivers/spi/spi_context.h:66]: @JordanYates When those changes will be ready? I can't see any MR with them.

**Comment by bbilas** [comment]
@ABOSTM @FRASTM @erwango gentle ping

**Review by FRASTM — APPROVED** [review]

**Review by ABOSTM — APPROVED** [review]

**Comment by erwango** [comment]
Set as DNM waiting for @tbursztyka approval

File: drivers/spi/spi_context.h
Comment (gmarull) [inline drivers/spi/spi_context.h:214]: Function defined in a header should be `static inline`

File: drivers/spi/spi_context.h
Comment (gmarull) [inline drivers/spi/spi_context.h:67]: this is duplicating `GPIO_DT_SPEC_GET_BY_IDX`, but with different field names, why not modify SPI to use gpio_dt_spec?

File: drivers/spi/spi_context.h
Comment (gmarull) [inline drivers/spi/spi_context.h:211]: why not modify SPI to use `gpio_dt_spec`?

File: drivers/spi/spi_context.h
Comment (gmarull) [inline drivers/spi/spi_context.h:223]: `err < 0` (essentially boolean expression)

File: drivers/spi/spi_context.h
Comment (gmarull) [inline drivers/spi/spi_context.h:243]: `spi_context_cs_configure` should be modified to return error code, this will hide the problem, e.g. if no logging is enabled.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h:67]: spi_cs_control struct does not use gpio_dt_spec so I've decided to not touch this and use it as it's and convert that later when it will be done. Take a look at the above conversation.

Comment (gmarull) [reply drivers/spi/spi_context.h:67]: I think it's better to _modernize_ SPI other than adding duplicate code on top of it.

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h:243]: I saw it but I thought that someone had some serious reason to ignore the return code from `gpio_pin_configure` function so I haven't changed that. I will fix it.

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h:67]: Sure, I just didn't want to duplicate the work that @JordanYates has done here https://github.com/csiro-wsn/zephyr/commits/210801_spi_cs

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h]: Fixed.

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h]: Fixed.

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h:67]: I'll change that when https://github.com/zephyrproject-rtos/zephyr/pull/39116 will be merged.

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h:211]: as above

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h]: Fixed.

**Review by mniestroj — APPROVED** [review]

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h]: Fixed.

File: drivers/spi/spi_context.h
Comment (bbilas) [reply drivers/spi/spi_context.h]: Fixed.

File: drivers/spi/spi_context.h
Comment (JordanYates) [inline drivers/spi/spi_context.h:199]: The naming of these functions is misleading to me, they are all `_configure` but they only configure in one direction, unlike the GPIO functions for example. Perhaps `initialize`?

File: drivers/spi/spi_ll_stm32.c
Comment (JordanYates) [inline drivers/spi/spi_ll_stm32.c:849]: If this logic makes sense generically, the rest of the SPI drivers should also be updated to do this.
If it does not make sense generally, the new functions probably shouldn't be in the global headers.
I don't think we want to introduce more family specific behaviors if we can help it (SPI CS lines are setup on STM, but not any others).

**Review by JordanYates — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_ll_stm32.c
Comment (bbilas) [reply drivers/spi/spi_ll_stm32.c:849]: The aim is to provide the same support for all supported drivers but currently I have only one device with stm32f4 that can be used for testing. Next week, I'll get some devboards with nRF52 chips so I'll add these changes for them too.

**Review by gmarull — COMMENTED** [review]
I assume all `spi_context_cs_configure` could now be deleted from drivers, right?

Since `spi_context_cs_configure_all` is called during init, it should also check for GPIO device readiness, and remove that check from `spi_ready()`. 

I wonder if there should also be another helper that calls `gpio_configure` with `GPIO_DISCONNECTED` (for PM purposes), but that's probably for another PR.

Just some questions after getting some more context.

**Comment by bbilas** [comment]
> I assume all `spi_context_cs_configure` could now be deleted from drivers, right?
> 
> Since `spi_context_cs_configure_all` is called during init, it should also check for GPIO device readiness, and remove that check from `spi_ready()`.
> 
I think so, we can do something like that
<pre>
static inline int spi_context_cs_configure_all(struct spi_context *ctx)
{
	int ret;
	const struct gpio_dt_spec *cs_gpio;

	for (cs_gpio = ctx->cs_gpios; cs_gpio < &ctx->cs_gpios[ctx->num_cs_gpios]; cs_gpio++) {
		if (!device_is_ready(cs_gpio->port) {
	                LOG_ERR("GPIO port %s is not ready", cs_gpio->port->name);
	                return -ENODEV;
		}

		/* Validate CS active levels are equivalent */
		__ASSERT(spi_context_cs_active_level(ctx) ==
			 (cs_gpio->gpio.dt_flags & GPIO_ACTIVE_LOW),
			 "Devicetree and spi_context CS levels are not equal");

		ret = gpio_pin_configure_dt(cs_gpio, GPIO_OUTPUT_INACTIVE);
		if (ret < 0) {
			return ret;
		}
	}

	return 0;
}
</pre>
> I wonder if there should also be another helper that calls `gpio_configure` with `GPIO_DISCONNECTED` (for PM purposes), but that's probably for another PR.
> 
> Just some questions after getting some more context.

I'm not familiar with power management in zephyr so it's hard for me to say something.

**Review by mniestroj — CHANGES_REQUESTED** [review]

**Review by mniestroj — APPROVED** [review]
LGTM

**Review by gmarull — APPROVED** [review]

**Comment by erwango** [comment]
@tbursztyka would you have time to review ?

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**
